src/JSInterop/Microsoft.JSInterop/src/DotNetObjectRefOfT.cs

```diff
@@ -0,0 +1,117 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.ComponentModel;
+using System.Text.Json.Serialization;
+
+namespace Microsoft.JSInterop
+{
+    /// <summary>
+    /// Wraps a JS interop argument, indicating that the value should not be serialized as JSON
+    /// but instead should be passed as a reference.
+    ///
+    /// To avoid leaking memory, the reference must later be disposed by JS code or by .NET code.
+    /// </summary>
+    /// <typeparam name="TValue">The type of the value to wrap.</typeparam>
+    public sealed class DotNetObjectRef<TValue> : IDotNetObjectRef, IDisposable where TValue : class
```

Comment:
Here's the interesting bit:

1) During serialization, we ignore the Value property and serialize the object id with a unique key: `{ "__dotnetObjectRef": 2334 }`
2) JavaScript looks for an object with the key and turns in to a `DotNetObjectRef` instance.
3) For a call in to .NET, when the serializer sets the `TrackingId` property, we lookup the object.
4) Additionally if the dispatcher notices a top-level argument that's shaped like DotNetObjectRef, but the receiver is not an instance of this type, we throw. At least for the most common cases where a user could mess up, we produce a decent error.